### PR TITLE
[low-power] add RCP support for CSL

### PR DESCRIPTION
**Pull request by brianljt**

### CODE REVIEW

File: include/openthread/platform/radio.h [hunk 1]
```diff
@@ -513,6 +514,16 @@ void otPlatRadioSetMacFrameCounter(otInstance *aInstance, uint32_t aMacFrameCoun
  */
 uint64_t otPlatRadioGetNow(otInstance *aInstance);
 
+/**
+ * Get the bus speed between the host and the radio chip.
+ *
+ * @param[in]   aInstance    A pointer to an OpenThread instance.
+ *
+ * @returns The bus speed between host and the radio chip. 0 when platform does not use bus.
```
Comment (edmont) [inline include/openthread/platform/radio.h:522]: Units?

File: src/core/mac/sub_mac.cpp [hunk 2]
```diff
@@ -724,6 +736,24 @@ bool SubMac::ShouldHandleEnergyScan(void) const
     return swEnergyScan;
 }
 
+bool SubMac::ShouldHandleTransmitDelay(void) const
+{
+    bool swTxSecurity = true;
```
Comment (Irving-cl) [inline src/core/mac/sub_mac.cpp:741]:
```suggestion
    bool swTxDelay = true;
```

File: src/core/thread/csl_tx_scheduler.cpp [hunk 3]
```diff
@@ -66,6 +66,18 @@ CslTxScheduler::CslTxScheduler(Instance &aInstance)
...
+}
+
+void CslTxScheduler::InitFrameRequestAhead(void)
+{
+    uint32_t busSpeedHz = otPlatRadioGetBusSpeed(&GetInstance());
+    // longest frame on bus is 127 bytes with some metadata, use 150 bytes for bus Tx time estimation
+    uint32_t busTxTimeUs = ((busSpeedHz == 0) ? 0 : (150 * 8 * 1000000 + busSpeedHz - 1) / busSpeedHz);
+
+    // Use ceiling to get next closest integer
+    mCslFrameRequestAhead =
```
Comment (Irving-cl) [inline src/core/thread/csl_tx_scheduler.cpp:79]: For non-RCP, set it to a constant for `Mac` handling delay?

File: src/posix/platform/openthread-core-posix-config.h [hunk 4]
```diff
@@ -104,6 +104,16 @@
 #define OPENTHREAD_CONFIG_PLATFORM_RADIO_COEX_ENABLE 1
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_PLATFORM_USEC_TIMER_ENABLE
+ *
+ * Define to 1 if you want to enable microsecond backoff timer implemented in platform.
```
Comment (Irving-cl) [inline src/posix/platform/openthread-core-posix-config.h:110]:
```suggestion
 * Define to 1 to enable microsecond backoff timer implemented in platform.
```

File: src/core/thread/csl_tx_scheduler.cpp (hunk 3 above)
Comment (brianljt) [reply src/core/thread/csl_tx_scheduler.cpp:79]: For non-RCP, the busSpeedHz will be 0 so that it will not add any bus transmission time to it. It will just use the `OPENTHREAD_CONFIG_MAC_CSL_REQUEST_AHEAD_THRESHOLD_US`.

Comment (Irving-cl) [reply src/core/thread/csl_tx_scheduler.cpp:79]: OK.

File: src/core/config/mac.h [hunk 5]
```diff
@@ -139,6 +139,16 @@
 #define OPENTHREAD_CONFIG_MAC_MAX_TX_ATTEMPTS_INDIRECT_POLLS 4
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_MAC_CSL_REQUEST_AHEAD_THRESHOLD_US
+ *
+ * Define how many microseconds ahead should MAC deliver CSL frame to SubMac.
```
Comment (edmont) [inline src/core/config/mac.h:145]: Where does this number come from? How can the user calculate it for their platform?

**Comment by edmont** [comment]
@brianljt won't this PR consider the `OT_RADIO_CAPS_RECEIVE_TIMING` case?

File: src/core/config/mac.h (hunk 5 above)
Comment (Irving-cl) [reply src/core/config/mac.h:145]: This is not a platform-specific value. It's used to minimize the exception cases that MAC is somehow delayed to pass CSL frame to SubMac, and SubMac has to send the frame in next slot. So theoretically it cannot 100% ensure this. It's an experienced value.

**Comment by brianljt** [comment]
> @brianljt won't this PR consider the `OT_RADIO_CAPS_RECEIVE_TIMING` case?

This PR is only for enabling CSL for RCP mode. So it does not need `OT_RADIO_CAPS_RECEIVE_TIMING` yet.

**Review by Irving-cl — APPROVED** [review]
LGTM👍

File: include/openthread/platform/radio.h [hunk 6]
```diff
@@ -513,6 +514,17 @@ void otPlatRadioSetMacFrameCounter(otInstance *aInstance, uint32_t aMacFrameCoun
  */
 uint64_t otPlatRadioGetNow(otInstance *aInstance);
 
+/**
+ * Get the bus speed in Hz(bps) between the host and the radio chip.
```
Comment (bukepo) [inline include/openthread/platform/radio.h:518]: Use a unified unit(bps)?

File: .github/workflows/simulation-1.2.yml [hunk 7]
```diff
@@ -92,3 +92,31 @@ jobs:
...
+    runs-on: ubuntu-18.04
+    env:
+      COVERAGE: 1
+      THREAD_VERSION: 1.2
+      VIRTUAL_TIME: 0
+    steps:
+    - uses: actions/checkout@v2
+    - name: Bootstrap
+      run: |
+        OT_OPTIONS=-DOT_READLINE=OFF sudo apt-get --no-install-recommends install -y expect ninja-build
```
Comment (bukepo) [inline .github/workflows/simulation-1.2.yml:106]:
```suggestion
        sudo apt-get --no-install-recommends install -y expect ninja-build
```

File: .github/workflows/simulation-1.2.yml [hunk 8]
```diff
@@ -92,3 +92,31 @@ jobs:
...
+    - name: Run RCP Mode
+      run: |
+        OT_OPTIONS=-DOT_READLINE=OFF NODE_MODE=rcp ./script/test clean build expect
```
Comment (bukepo) [inline .github/workflows/simulation-1.2.yml:109]:
```suggestion
        OT_OPTIONS=-DOT_READLINE=OFF NODE_MODE=rcp ./script/test build expect
```

File: src/lib/spinel/radio_spinel_impl.hpp [hunk 9]
```diff
@@ -317,10 +320,15 @@ otError RadioSpinel<InterfaceType, ProcessContextType>::CheckRadioCapabilities(v
...
     {
-        otLogCritPlat("RCP does not support required capabilities: ack-timeout:%s, tx-retries:%s, CSMA-backoff:%s",
-                      (mRadioCaps & OT_RADIO_CAPS_ACK_TIMEOUT) ? "yes" : "no",
-                      (mRadioCaps & OT_RADIO_CAPS_TRANSMIT_RETRIES) ? "yes" : "no",
-                      (mRadioCaps & OT_RADIO_CAPS_CSMA_BACKOFF) ? "yes" : "no");
+        otRadioCaps missingCaps = (mRadioCaps & kRequiredRadioCaps) ^ kRequiredRadioCaps;
+
+        otLogCritPlat("RCP capabilities check failed: ack-timeout:%s, tx-retries:%s, CSMA-backoff:%s, tx-security:%s, "
+                      "tx-timing:%s",
+                      (missingCaps & OT_RADIO_CAPS_ACK_TIMEOUT) ? "failed" : "pass",
```
Comment (bukepo) [inline src/lib/spinel/radio_spinel_impl.hpp:327]: Is this change necessary?

**Comment by edmont** [comment]
@brianljt, one question, how is the `otPlatRadioTransmit()` implementation going to decide whether or not to apply the delayed transmission? Based on the Radio Caps?

File: src/lib/spinel/radio_spinel_impl.hpp (hunk 9 above)
Comment (brianljt) [reply src/lib/spinel/radio_spinel_impl.hpp:327]: If we use the old way. Newly added capabilities are only needed in `OPENTHREAD_CONFIG_MAC_CSL_TRANSMITTER_ENABLE`, then we have to add #if defines inside `otLogCritPlat` which will trigger macos-clang check error.

Change it this way will make it more clear that which capability is missing.

**Comment by brianljt** [comment]
> @brianljt, one question, how is the `otPlatRadioTransmit()` implementation going to decide whether or not to apply the delayed transmission? Based on the Radio Caps?

If `otPlatRadioTransmit()` supports delayed transmission, Radio should mark `OT_RADIO_CAPS_TRANSMIT_TIMING` in `otPlatRadioGetCaps`. With that `SubMac` will skip the delay in `StartCsmaBackoff` --> `ShouldHandleTransmitDelay` and leave it to Radio layer to handle the transmit delay.

**Comment by edmont** [comment]
@brianljt but I mean, how does the radio know a specific frame need to be delayed or not?

**Comment by brianljt** [comment]
> @brianljt but I mean, how does the radio know a specific frame need to be delayed or not?

If the frame is a CSL transmission, it will have `mTransmitFrame.mInfo.mTxInfo.mPeriod` and `mTransmitFrame.mInfo.mTxInfo.mPhase` set. So radio found `mTransmitFrame.mInfo.mTxInfo.mPeriod != 0`, it will get its current time and send the frame at  `mTransmitFrame.mInfo.mTxInfo.mPhase`.

BTW, we have fixed one issue in https://github.com/openthread/openthread/pull/5449#issuecomment-685329491 . Now the transmit timing on Nordic DK is quite good. The window can be set small enough.

File: src/lib/spinel/radio_spinel_impl.hpp (hunk 9 above)
Comment (bukepo) [reply src/lib/spinel/radio_spinel_impl.hpp:327]: It makes sense. How about only display missing features:
```cpp
otLogCritPlat("RCP does not support required capabilities:%s%s%s%s",
(missingCaps & OT_RADIO_CAPS_ACK_TIMEOUT) ? " ack-timeout" : "",
(missingCaps & OT_RADIO_CAPS_TRANSMIT_RETRIES) ? " tx-retries" : "",
....
```

Comment (brianljt) [reply src/lib/spinel/radio_spinel_impl.hpp]: Sounds good. Code updated.

**Review by bukepo — APPROVED** [review]
LGTM:+1:

**Review by abtink — APPROVED** [review]
Thanks @brianljt . Nicely done. LGTM.

**Comment by brianljt** [comment]
@edmont do you have any other suggestions for this PR?

File: include/openthread/platform/radio.h [hunk 10]
```diff
@@ -516,6 +517,17 @@ void otPlatRadioSetMacFrameCounter(otInstance *aInstance, uint32_t aMacFrameCoun
...
```
Comment (jwhui) [inline include/openthread/platform/radio.h:521]:
```suggestion
 * Get the bus speed in bits/second between the host and the radio chip.
```

File: include/openthread/platform/radio.h [hunk 11]
```diff
@@ -516,6 +517,17 @@ void otPlatRadioSetMacFrameCounter(otInstance *aInstance, uint32_t aMacFrameCoun
...
+ *
+ * @param[in]   aInstance    A pointer to an OpenThread instance.
+ *
+ * @returns The bus speed in bps between the host and the radio chip.
```
Comment (jwhui) [inline include/openthread/platform/radio.h:525]:
```suggestion
 * @returns The bus speed in bits/second between the host and the radio chip.
```

File: include/openthread/platform/radio.h [hunk 12]
```diff
@@ -516,6 +517,17 @@ void otPlatRadioSetMacFrameCounter(otInstance *aInstance, uint32_t aMacFrameCoun
...
+ *          Return 0 when platform does not use bus.
```
Comment (jwhui) [inline include/openthread/platform/radio.h:526]: What does it mean to not use a bus?

File: src/core/config/mac.h [hunk 13]
```diff
@@ -139,6 +139,16 @@
 #define OPENTHREAD_CONFIG_MAC_MAX_TX_ATTEMPTS_INDIRECT_POLLS 4
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_MAC_CSL_REQUEST_AHEAD_THRESHOLD_US
```
Comment (jwhui) [inline src/core/config/mac.h:143]: Is this really a "threshold"? Is this really just trying to target when the MAC will deliver the CSL frame to sub-MAC?

```suggestion
 * @def OPENTHREAD_CONFIG_MAC_CSL_REQUEST_AHEAD_US
```

File: src/core/config/mac.h [hunk 14]
```diff
@@ -295,6 +305,16 @@
 #define OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_SECURITY_ENABLE 0
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_TIMING_ENABLE
+ *
+ * Define to 1 to enable software transmission delay logic.
```
Comment (jwhui) [inline src/core/config/mac.h:311]:
```suggestion
 * Define to 1 to enable software transmission target time logic.
```

File: src/core/mac/sub_mac.cpp [hunk 15]
```diff
@@ -341,18 +345,26 @@ void SubMac::StartCsmaBackoff(void)
...
-        uint32_t phaseDesired = mTransmitFrame.mInfo.mTxInfo.mPhase;
-
         SetState(kStateCslTransmit);
 
-        if (phaseNow < phaseDesired)
-        {
-            mTimer.Start((phaseDesired - phaseNow) * kUsPerTenSymbols);
-        }
-        else if (phaseNow > phaseDesired)
+        if (ShouldHandleTransmitDelay())
```
Comment (jwhui) [inline src/core/mac/sub_mac.cpp:350]: Is "target time" more descriptive?
```suggestion
        if (ShouldHandleTransmitTargetTime())
```

File: src/lib/spinel/radio_spinel.hpp [hunk 16]
```diff
@@ -669,6 +669,14 @@ template <typename InterfaceType, typename ProcessContextType> class RadioSpinel
      */
     uint64_t GetNow(void);
 
+    /**
+     * This method get bus speed between the host and the radio.
+     *
+     * @returns   bus speed in bps.
```
Comment (jwhui) [inline src/lib/spinel/radio_spinel.hpp:675]:
```suggestion
     * @returns   Bus speed in bits/second.
```

File: src/lib/spinel/radio_spinel.hpp [hunk 17]
```diff
@@ -669,6 +669,14 @@ template <typename InterfaceType, typename ProcessContextType> class RadioSpinel
      */
     uint64_t GetNow(void);
 
+    /**
+     * This method get bus speed between the host and the radio.
```
Comment (jwhui) [inline src/lib/spinel/radio_spinel.hpp:673]:
```suggestion
     * This method returns the bus speed between the host and the radio.
```

File: src/posix/platform/hdlc_interface.hpp [hunk 18]
```diff
@@ -148,6 +148,14 @@ class HdlcInterface
     void Process(const VirtualTimeEvent &aEvent) { Decode(aEvent.mData, aEvent.mDataLength); }
 #endif
 
+    /**
+     * This method get bus speed between the host and the radio.
```
Comment (jwhui) [inline src/posix/platform/hdlc_interface.hpp:152]:
```suggestion
     * This method returns the bus speed between the host and the radio.
```

File: src/posix/platform/hdlc_interface.hpp [hunk 19]
```diff
@@ -148,6 +148,14 @@ class HdlcInterface
...
+     *
+     * @returns   bus speed in bps.
```
Comment (jwhui) [inline src/posix/platform/hdlc_interface.hpp:154]:
```suggestion
     * @returns   Bus speed in bits/second.
```

File: src/posix/platform/spi_interface.hpp [hunk 20]
```diff
@@ -138,6 +138,14 @@ class SpiInterface
      */
     void Process(const RadioProcessContext &aContext);
 
+    /**
+     * This method get bus speed between the host and the radio.
```
Comment (jwhui) [inline src/posix/platform/spi_interface.hpp:142]:
```suggestion
     * This method returns the bus speed between the host and the radio.
```

File: src/posix/platform/spi_interface.hpp [hunk 21]
```diff
@@ -138,6 +138,14 @@ class SpiInterface
...
+     *
+     * @returns   bus speed in bps.
```
Comment (jwhui) [inline src/posix/platform/spi_interface.hpp:144]:
```suggestion
     * @returns   Bus speed in bits/second.
```

File: examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h [hunk 22]
```diff
@@ -150,6 +150,16 @@
 #define OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_SECURITY_ENABLE (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_TIMING_ENABLE
+ *
+ * Define to 1 to enable software transmission delay logic.
```
Comment (jwhui) [inline examples/platforms/nrf528xx/nrf52811/openthread-core-nrf52811-config.h:156]:
```suggestion
 * Define to 1 to enable software transmission target time logic.
```

File: examples/platforms/nrf528xx/nrf52833/openthread-core-nrf52833-config.h [hunk 23]
```diff
@@ -172,6 +172,16 @@
 #define OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_SECURITY_ENABLE (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_TIMING_ENABLE
+ *
+ * Define to 1 to enable software transmission delay logic.
```
Comment (jwhui) [inline examples/platforms/nrf528xx/nrf52833/openthread-core-nrf52833-config.h:178]:
```suggestion
 * Define to 1 to enable software transmission target time logic.
```

File: examples/platforms/nrf528xx/nrf52840/openthread-core-nrf52840-config.h [hunk 24]
```diff
@@ -172,6 +172,16 @@
 #define OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_SECURITY_ENABLE (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
 #endif
 
+/**
+ * @def OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_TIMING_ENABLE
+ *
+ * Define to 1 to enable software transmission delay logic.
```
Comment (jwhui) [inline examples/platforms/nrf528xx/nrf52840/openthread-core-nrf52840-config.h:178]:
```suggestion
 * Define to 1 to enable software transmission target time logic.
```

File: examples/platforms/simulation/openthread-core-simulation-config.h [hunk 25]
```diff
@@ -117,6 +117,16 @@
 #ifndef OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_SECURITY_ENABLE
 #define OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_SECURITY_ENABLE (OPENTHREAD_CONFIG_THREAD_VERSION >= OT_THREAD_VERSION_1_2)
 #endif
+
+/**
+ * @def OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_TIMING_ENABLE
+ *
+ * Define to 1 to enable software transmission delay logic.
```
Comment (jwhui) [inline examples/platforms/simulation/openthread-core-simulation-config.h:124]:
```suggestion
 * Define to 1 to enable software transmission target time logic.
```

File: tests/toranj/openthread-core-toranj-config.h [hunk 26]
```diff
@@ -462,4 +462,12 @@
  *
  */
 #define OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_SECURITY_ENABLE 1
+
+/**
+ * @def OPENTHREAD_CONFIG_MAC_SOFTWARE_TX_TIMING_ENABLE
+ *
+ * Define to 1 to enable software transmission delay logic.
```
Comment (jwhui) [inline tests/toranj/openthread-core-toranj-config.h:469]:
```suggestion
 * Define to 1 to enable software transmission target time logic.
```

**Review by jwhui — APPROVED** [review]
